core/src/main/scala/kafka/server/KafkaRequestHandler.scala

```diff
@@ -53,25 +53,27 @@ object KafkaRequestHandler {
    * Wrap callback to schedule it on a request thread.
    * NOTE: this function must be called on a request thread.
    * @param fun Callback function to execute
+   * @param requestLocal The RequestLocal for the current request handler thread in case we need to call
+   *                     the callback function without queueing the callback request
    * @return Wrapped callback that would execute `fun` on a request thread
    */
-  def wrap[T](fun: T => Unit): T => Unit = {
+  def wrap[T](fun: (RequestLocal, T) => Unit, requestLocal: RequestLocal): T => Unit = {
     val requestChannel = threadRequestChannel.get()
     val currentRequest = threadCurrentRequest.get()
     if (requestChannel == null || currentRequest == null) {
       if (!bypassThreadCheck)
         throw new IllegalStateException("Attempted to reschedule to request handler thread from non-request handler thread.")
-      T => fun(T)
+      T => fun(requestLocal, T)
     } else {
       T => {
-        if (threadCurrentRequest.get() != null) {
-          // If the callback is actually executed on a request thread, we can directly execute
+        if (threadCurrentRequest.get() == currentRequest) {
```

Comment:
Ismael mentioned in https://github.com/apache/kafka/pull/9229#issuecomment-683352094 that thread locals are most useful when one doesn't control the code. So, I am wondering if we could get rid of the two ThreadLocal: `threadRequestChannel` and `threadCurrentRequest` in `KafkaRequestHandler` introduced in KAFKA-14561.  The reason for the former is to obtain requestChannel. We could simply pass in requestChannel to `ReplicaManager.appendRecords`. The reason for the latter is (1) to obtain currentRequest and (2) to make sure that the callback can be short-circuited if it's called on the same request handler thread. For (1), we could also pass in currentRequest to `ReplicaManager.appendRecords`. For (2), we could change the code such that `KafkaRequestHandler.wrap` is called only when the callback truly needs to be run from a different thread. Otherwise, we can just call the callback directly there.

Reply:
Hmmm. I'm not sure it makes sense to try to pass the request channel and current in to every method we want to do a callback for.

As for 2 and short circuiting... Are you suggesting I move the wrap call into AddPartitionsToTxnManager? If so, I need to pass the appendEntries method in there as well. That may be trickier with the typing.

Reply:
> Hmmm. I'm not sure it makes sense to try to pass the request channel and current in to every method we want to do a callback for.

Currently, the callback is only needed for produce request in ReplicaManager. If you look at `KafkaApis.handleProduceRequest`, we already pass in both request channel and current request to `ReplicaManager.appendRecords` through `sendResponseCallback`.

> As for 2 and short circuiting... Are you suggesting I move the wrap call into AddPartitionsToTxnManager? If so, I need to pass the appendEntries method in there as well. That may be trickier with the typing.

Yes. Currently, we already pass in `appendEntries` as a callback to `AddPartitionsToTxnManager`, right?

Reply:
We pass the wrapped method which is a simpler type.  `type AppendCallback = Map[TopicPartition, Errors] => Unit`

Reply:
As for just produce. Yes, this is the case now, but I think Artem was trying to create the wrap method as a "general" callback mechanism.

Reply:
Yeah, the hope is that we'd see more async patterns as we evolve Kafka, and the way it's currently implemented, we just say "execute callback on the _current_ thread pool", which is why the thread local is used -- this way we can implement a non-blocking wait-and-continue-on-current-thread-pool from any level without passing additional arguments through the whole stack.  The same functionality could be implemented on other thread pools, e.g. if the group coordinator has its own thread pool, we could implement the same primitive there as well.

Reply:
Thanks, Justine and Artem. Yes, passing around the context has it's own issues. The main thing with thread local is to make sure that we don't introduce any GC issues. Currently, it seems that we never remove `threadRequestChannel`. Since we allow dynamically changing the number of request handler threads, it's probably better to remove `threadRequestChannel` when the thread completes?

Reply:
Hmm. We assign the request channel when the thread is created. I assume that request channel will remain with the thread during its lifetime. Is that incorrect?

Reply:
True in the common case. I was wondering what happens if we dynamically reduce the number of request handler threads.

Reply:
I we reduce then I assume the thread just stops?

Reply:
I understand the issue. I have a commit below that clears the request local.